File: src/main/java/dark/leech/text/plugin/js/loader/JsApiSetup.java
package dark.leech.text.plugin.js.loader;

import org.mozilla.javascript.Context;
import org.mozilla.javascript.Scriptable;
import org.mozilla.javascript.ScriptableObject;

import dark.leech.text.action.Log;
import dark.leech.text.plugin.js.api.Engine;
import dark.leech.text.plugin.js.api.Html;
import dark.leech.text.plugin.js.api.Http;
import dark.leech.text.plugin.js.api.Json;
import dark.leech.text.plugin.js.api.LocalStorage;
import dark.leech.text.plugin.js.api.UserAgent;

/**
 * Shared vBook API setup helper for Rhino JavaScript loaders. Extracts duplicate setup logic from
 * TextLoader, DetailLoader, and ListLoader.
 */
public final class JsApiSetup {

    private JsApiSetup() {
        // Utility class - prevent instantiation
    }

    /**
     * Setup vBook API bindings in the provided scope. Registers BASE_URL, Html, Response, fetch,
     * load, and sleep functions.
     *
     * @param ctx The Rhino context
     * @param scope The Rhino scope
     * @param baseUrl The base URL for the plugin source
     * @param targetUrl The target URL being processed
     * @param pluginSource The plugin source directory (optional, for load() function)
     */
    public static void setup(
            Context ctx, Scriptable scope, String baseUrl, String targetUrl, String pluginSource) {
        // Set BASE_URL variable
        ScriptableObject.putProperty(scope, "BASE_URL", baseUrl);

        // Match vBooks runtime behavior: expose Java String as JS primitive string
        // so JS String methods (e.g. split('?')) follow JavaScript semantics.
        ctx.getWrapFactory().setJavaPrimitiveWrap(false);

        // Html object - Rhino automatically exposes public methods
        Html htmlApi = new Html(ctx, scope);
        ScriptableObject.putProperty(scope, "Html", htmlApi);
        ctx.getWrapFactory().setJavaPrimitiveWrap(false);

        // Http class - expose for static method calls like Http.get(), Http.post()
        Http httpInstance = new Http(ctx, scope);
        ScriptableObject.putProperty(scope, "Http", httpInstance);
        ctx.getWrapFactory().setJavaPrimitiveWrap(false);

        // Json class - expose for JSON parsing (vBook compatibility)
        Json jsonApi = new Json(ctx, scope);
        ScriptableObject.putProperty(scope, "Json", jsonApi);
        ctx.getWrapFactory().setJavaPrimitiveWrap(false);

        // UserAgent class - expose for platform-specific user agents
        UserAgent userAgentApi = new UserAgent(ctx, scope);
        ScriptableObject.putProperty(scope, "UserAgent", userAgentApi);
        ctx.getWrapFactory().setJavaPrimitiveWrap(false);

        // Engine class - expose for browser automation
        Engine engineApi = new Engine(ctx, scope);
        ScriptableObject.putProperty(scope, "Engine", engineApi);
        ctx.getWrapFactory().setJavaPrimitiveWrap(false);

        // Response object
        Response responseObj = new Response(ctx, scope);
        ScriptableObject.putProperty(scope, "Response", responseObj);

        // Fetch function - returns Http object for chaining
        Object fetchFunc =
                new org.mozilla.javascript.BaseFunction() {
                    @Override
                    public Object call(
                            org.mozilla.javascript.Context cx,
                            Scriptable scope,
                            Scriptable thisObj,
                            Object[] args) {
                        // Defensive: use Context.toString() to handle any Rhino object type
                        String requestUrl;
                        if (args.length > 0
                                && args[0] != null
                                && args[0] != org.mozilla.javascript.Undefined.instance) {
                            requestUrl = org.mozilla.javascript.Context.toString(args[0]);
                        } else {
                            requestUrl = targetUrl;
                        }
                        Http http = new Http(cx, scope);
                        http.request(requestUrl);
                        return http;
                    }

                    @Override
                    public Object getDefaultValue(java.lang.Class<?> typeHint) {
                        // Safe toString() implementation for plugin compatibility
                        return "function fetch() { [native code] }";
                    }
                };
        ScriptableObject.putProperty(scope, "fetch", fetchFunc);
        ctx.getWrapFactory().setJavaPrimitiveWrap(false);

        // Load function - load JavaScript files from plugin directory
        Object loadFunc =
                new org.mozilla.javascript.BaseFunction() {
                    @Override
                    public Object call(
                            org.mozilla.javascript.Context cx,
                            Scriptable scope,
                            Scriptable thisObj,
                            Object[] args) {
                        String fileName =
                                (args.length > 0 && args[0] != null)
                                        ? JSResponse.getString(args[0])
                                        : "";

                        if (fileName.isEmpty()) {
                            Log.add("[load()] File name is empty");
                            return null;
                        }

                        // Security: Validate file path
                        if (!isValidFileName(fileName)) {
                            Log.add("[load()] Invalid file name: " + fileName);
                            return null;
                        }

                        // Load file from plugin directory
                        if (pluginSource == null || pluginSource.isEmpty()) {
                            Log.add("[load()] Plugin source directory not set");
                            return null;
                        }

                        try {
                            java.io.File file = new java.io.File(pluginSource, "src/" + fileName);

                            // Security: Check if file exists and is within plugin directory
                            if (!file.exists() || !file.isFile()) {
                                Log.add("[load()] File not found: " + file.getPath());
                                return null;
                            }

                            // Security: Check file path doesn't escape plugin directory
                            java.io.File pluginDir = new java.io.File(pluginSource, "src");
                            if (!file.getCanonicalPath().startsWith(pluginDir.getCanonicalPath())) {
                                Log.add("[load()] Path traversal detected: " + fileName);
                                return null;
                            }

                            // Security: Check file size (limit to 100KB)
                            long fileSize = file.length();
                            if (fileSize > 100 * 1024) {
                                Log.add("[load()] File too large: " + fileSize + " bytes");
                                return null;
                            }

                            // Read file content
                            String content =
                                    new String(java.nio.file.Files.readAllBytes(file.toPath()));

                            // Execute script in current context
                            return cx.evaluateString(scope, content, fileName, 1, null);

                        } catch (Exception e) {
                            Log.add("[load()] Failed to load file: " + e.getMessage());
                            return null;
                        }
                    }

                    @Override
                    public Object getDefaultValue(java.lang.Class<?> typeHint) {
                        // Safe toString() implementation for plugin compatibility
                        return "function load() { [native code] }";
                    }
                };
        ScriptableObject.putProperty(scope, "load", loadFunc);
        ctx.getWrapFactory().setJavaPrimitiveWrap(false);

        // Sleep function (optional - only used by ListLoader)
        Object sleepFunc =
                new org.mozilla.javascript.BaseFunction() {
                    @Override
                    public Object call(
                            org.mozilla.javascript.Context cx,
                            Scriptable scope,
                            Scriptable thisObj,
                            Object[] args) {
                        int millis =
                                (args.length > 0 && args[0] instanceof Number)
                                        ? ((Number) args[0]).intValue()
                                        : 0;
                        try {
                            Thread.sleep(millis);
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                        }
                        return null;
                    }

                    @Override
                    public Object getDefaultValue(java.lang.Class<?> typeHint) {
                        // Safe toString() implementation for plugin compatibility
                        return "function sleep() { [native code] }";
                    }
                };
        ScriptableObject.putProperty(scope, "sleep", sleepFunc);
        ctx.getWrapFactory().setJavaPrimitiveWrap(false);

        // Console API for plugin logging compatibility
        // Match trusted vBooks behavior: expose same logger object as both Console and console
        Object consoleApi = Context.javaToJS(new ConsoleApi(), scope);
        ScriptableObject.putProperty(scope, "Console", consoleApi);
        ScriptableObject.putProperty(scope, "console", consoleApi);
        ctx.getWrapFactory().setJavaPrimitiveWrap(false);

        // LocalStorage API for plugin persistence
        LocalStorage localStorageApi =
                new LocalStorage(ctx, scope, pluginSource != null ? pluginSource : "default");
        ScriptableObject.putProperty(scope, "localStorage", localStorageApi);
        ctx.getWrapFactory().setJavaPrimitiveWrap(false);
    }

    /**
     * Extract base URL from a full URL. Returns protocol://host[:port]
     *
     * @param url The URL to extract from
     * @return The base URL
     */
    public static String extractBaseUrl(String url) {
        try {
            java.net.URL parsed = new java.net.URL(url);
            String protocol = parsed.getProtocol();
            String host = parsed.getHost();
            int port = parsed.getPort();
            if (port == -1) {
                return protocol + "://" + host;
            }
            return protocol + "://" + host + ":" + port;
        } catch (Exception e) {
            return url;
        }
    }

    /**
     * Validate file name for security. Prevents path traversal and ensures only .js files are
     * allowed.
     *
     * @param fileName The file name to validate
     * @return true if valid, false otherwise
     */
    private static boolean isValidFileName(String fileName) {
        if (fileName == null || fileName.isEmpty()) {
            return false;
        }

        // No path traversal
        if (fileName.contains("..") || fileName.contains("/") || fileName.contains("\\")) {
            return false;
        }

        // No absolute paths
        if (fileName.startsWith("/") || fileName.startsWith("\\")) {
            return false;
        }

        // Must end with .js
        if (!fileName.endsWith(".js")) {
            return false;
        }

        // No special characters that could cause issues
        if (fileName.matches(".*[<>:\"|?*].*")) {
            return false;
        }

        return true;
    }
}

File: engine/src/sharedJvmMain/kotlin/dev/haipham22/leechtext/plugin/js/loader/JsApiSetup.kt
package dev.haipham22.leechtext.plugin.js.loader

import dev.haipham22.leechtext.log.EngineLogger
import dev.haipham22.leechtext.plugin.js.api.Engine
import dev.haipham22.leechtext.plugin.js.api.Html
import dev.haipham22.leechtext.plugin.js.api.Http
import dev.haipham22.leechtext.plugin.js.api.JsValueBridge
import dev.haipham22.leechtext.plugin.js.api.Json
import dev.haipham22.leechtext.plugin.js.api.LocalStorage
import dev.haipham22.leechtext.plugin.js.api.UserAgent
import dev.haipham22.leechtext.util.parseBaseUrl
import org.mozilla.javascript.BaseFunction
import org.mozilla.javascript.Context
import org.mozilla.javascript.NativeObject
import org.mozilla.javascript.Scriptable
import org.mozilla.javascript.ScriptableObject
import org.mozilla.javascript.Undefined
import java.io.File
import java.nio.file.Files

/**
 * Setup vBook API bindings vào Rhino scope (port từ loader/JsApiSetup.java; P5.2c: bind
 * instance API common — Rhino wrap NativeJavaObject như cũ).
 * Đăng ký: BASE_URL, Html, Http, Json, UserAgent, Engine, Response, fetch, load, sleep,
 * console, localStorage, localConfig.
 */
object JsApiSetup {
    @Suppress("LongParameterList") // tham số log DI + params nghiệp vụ — bundle làm code khó đọc
    fun setup(
        ctx: Context,
        scope: Scriptable,
        baseUrl: String?,
        targetUrl: String?,
        pluginSource: String?,
        log: EngineLogger,
    ) {
        setup(ctx, scope, baseUrl, targetUrl, pluginSource, null, log)
    }

    @Suppress("LongParameterList") // tham số log DI + params nghiệp vụ — bundle làm code khó đọc
    fun setup(
        ctx: Context,
        scope: Scriptable,
        baseUrl: String?,
        targetUrl: String?,
        pluginSource: String?,
        pluginConfig: Map<String, String>?,
        log: EngineLogger,
    ) = setup(ctx, scope, baseUrl, targetUrl, pluginSource, pluginConfig, null, log)

    @Suppress("LongParameterList") // tham số log DI + params nghiệp vụ — bundle làm code khó đọc
    fun setup(
        ctx: Context,
        scope: Scriptable,
        baseUrl: String?,
        targetUrl: String?,
        pluginSource: String?,
        pluginConfig: Map<String, String>?,
        extraScripts: Map<String, String>?,
        log: EngineLogger,
    ) {
        injectBaseUrl(ctx, scope, baseUrl, extraScripts)
        injectPluginConfig(ctx, scope, pluginConfig)
        registerLocalConfig(ctx, scope, pluginConfig)

        ScriptableObject.putProperty(scope, "Html", Html(log))
        ctx.wrapFactory.isJavaPrimitiveWrap = false
        ScriptableObject.putProperty(scope, "Http", Http(log))
        ctx.wrapFactory.isJavaPrimitiveWrap = false
        ScriptableObject.putProperty(scope, "Json", Json(log))
        ctx.wrapFactory.isJavaPrimitiveWrap = false
        ScriptableObject.putProperty(scope, "UserAgent", UserAgent())
        ctx.wrapFactory.isJavaPrimitiveWrap = false
        ScriptableObject.putProperty(scope, "Engine", Engine())
        ctx.wrapFactory.isJavaPrimitiveWrap = false
        ScriptableObject.putProperty(scope, "Response", Response())

        // fetch() — trả Http object cho chaining
        ScriptableObject.putProperty(scope, "fetch", createFetchFunc(targetUrl, log))
        ctx.wrapFactory.isJavaPrimitiveWrap = false

        // load() — đọc JS file từ plugin dir (có path-traversal guard, giới hạn 100KB)
        ScriptableObject.putProperty(scope, "load", createLoadFunc(pluginSource, extraScripts, log))
        ctx.wrapFactory.isJavaPrimitiveWrap = false

        // sleep(ms)
        ScriptableObject.putProperty(scope, "sleep", createSleepFunc())
        ctx.wrapFactory.isJavaPrimitiveWrap = false

        // Console API — expose cùng object cho Console + console (match vBooks)
        val consoleApi = Context.javaToJS(ConsoleApi(log), scope)
        ScriptableObject.putProperty(scope, "Console", consoleApi)
        ScriptableObject.putProperty(scope, "console", consoleApi)
        ctx.wrapFactory.isJavaPrimitiveWrap = false

        // LocalStorage
        ScriptableObject.putProperty(
            scope,
            "localStorage",
            LocalStorage(log, pluginSource ?: "default"),
        )
        ctx.wrapFactory.isJavaPrimitiveWrap = false
    }

    /** Inject BASE_URL vào scope làm JS string primitive (match vBooks). */
    private fun injectBaseUrl(
        ctx: Context,
        scope: Scriptable,
        baseUrl: String?,
        extraScripts: Map<String, String>?,
    ) {
        // Dùng evaluateString để guarantee JS string type — putProperty với Java String
        // đôi khi bị wrap làm NativeJavaObject, stringify ra rác (": NOT_FOUND").
        // Plugin có config.js (extraScripts) thì BỎ inject — config tự const BASE_URL
        // đúng gốc (plugin.source có thể lệch path, vd /reader nhân đôi URL).
        val hasConfigScript = extraScripts?.containsKey("config.js") == true
        if (baseUrl != null && !hasConfigScript) {
            val escaped =
                baseUrl
                    .replace("\\", "\\\\")
                    .replace("\"", "\\\"")
                    .replace("\n", "\\n")
                    .replace("\r", "\\r")
            ctx.evaluateString(scope, "var BASE_URL = \"$escaped\";", "baseUrl", 1, null)
        }
        ctx.wrapFactory.isJavaPrimitiveWrap = false
    }

    /** Inject plugin config keys làm JS constants (vBook extension-api). */
    private fun injectPluginConfig(
        ctx: Context,
        scope: Scriptable,
        pluginConfig: Map<String, String>?,
    ) {
        // vd config DOMAIN default "https://x.com" → const DOMAIN = "https://x.com"
        pluginConfig?.forEach { (key, value) ->
            if (key.isNotEmpty() && key.matches(Regex("[A-Za-z_][A-Za-z0-9_]*"))) {
                ScriptableObject.putProperty(scope, key, value)
                ctx.wrapFactory.isJavaPrimitiveWrap = false
            }
        }
    }

    /** localConfig API — getItem/setItem cho dynamic access (vBook extension-api). */
    private fun registerLocalConfig(
        ctx: Context,
        scope: Scriptable,
        pluginConfig: Map<String, String>?,
    ) {
        val localConfigObj =
            object : BaseFunction() {
                override fun call(
                    cx: Context,
                    sc: Scriptable,
                    thisObj: Scriptable,
                    args: Array<Any?>,
                ): Any? {
                    return null // getItem/setItem handled qua get
                }

                override fun get(
                    name: String,
                    start: Scriptable,
                ): Any {
                    if (name == "getItem") {
                        return object : BaseFunction() {
                            override fun call(
                                cx: Context,
                                sc: Scriptable,
                                th: Scriptable,
                                args: Array<Any?>,
                            ): Any? {
                                val key = args.getOrNull(0)?.toString() ?: return null
                                return pluginConfig?.get(key)
                            }
                        }
                    }
                    if (name == "setItem") {
                        return object : BaseFunction() {
                            override fun call(
                                cx: Context,
                                sc: Scriptable,
                                th: Scriptable,
                                args: Array<Any?>,
                            ): Any? {
                                // Read-only trong engine — không persist config change từ script
                                return null
                            }
                        }
                    }
                    return super.get(name, start)
                }
            }
        ScriptableObject.putProperty(scope, "localConfig", localConfigObj)
        ctx.wrapFactory.isJavaPrimitiveWrap = false
    }

    /** fetch() — trả Http object cho chaining, hỗ trợ option {method, body, headers, queries}. */
    private fun createFetchFunc(
        targetUrl: String?,
        log: EngineLogger,
    ): BaseFunction = object : BaseFunction() {
        override fun call(
            cx: Context,
            scope: Scriptable,
            thisObj: Scriptable,
            args: Array<Any?>,
        ): Any {
            var requestUrl = resolveRequestUrl(args, targetUrl)
            val opts = args.getOrNull(1) as? NativeObject
            val hasOptions = opts != null && listOf("method", "body", "headers", "queries").any { opts.has(it, opts) }
            if (opts != null && !hasOptions) {
                // Hành vi cũ: object trần (không key option) = query string
                requestUrl = appendQueryParams(requestUrl, opts)
            }
            log.add("[fetch()] $requestUrl")
            val http = Http(log).request(requestUrl)
            if (opts != null && hasOptions) {
                applyFetchOptions(http, opts)
            }
            return http
        }

        override fun getDefaultValue(typeHint: Class<*>?): Any = "function fetch() { [native code] }"
    }

    /** vBook extension-api: fetch(url, {method, body, headers, queries}) — apply option lên Http. */
    private fun applyFetchOptions(
        http: Http,
        opts: NativeObject,
    ) {
        val queries = opts.get("queries", opts)
        if (queries is Map<*, *>) http.queries(queries)
        val method = opts.get("method", opts)
        if (method is String) http.method(method)
        val headers = opts.get("headers", opts)
        if (headers is Map<*, *>) {
            for ((k, v) in headers) {
                if (k is String) http.header(k, v?.toString() ?: "")
            }
        }
        val body = opts.get("body", opts)
        when (body) {
            // object → form-encoded (vd ajax POST)
            is Map<*, *> -> http.form(body)

            is String -> http.body(body)
        }
    }

    /** Xác định URL request: arg đầu tiên hoặc fallback targetUrl. */
    private fun resolveRequestUrl(
        args: Array<Any?>,
        targetUrl: String?,
    ): String = if (args.isNotEmpty() && args[0] != null && args[0] !== Undefined.instance) {
        Context.toString(args[0])
    } else {
        targetUrl ?: ""
    }

    /** Append query string từ option {queries: {k: v}} vào URL. */
    private fun appendQueryParams(
        requestUrl: String,
        opts: NativeObject,
    ): String {
        val queries = opts.get("queries", opts)
        if (queries is NativeObject) {
            val pairs =
                queries.keys.mapNotNull { key ->
                    val k = key as? String ?: return@mapNotNull null
                    val v = queries.get(k, queries) ?: return@mapNotNull null
                    "$k=" +
                        java.net.URLEncoder.encode(
                            Context.toString(v),
                            Charsets.UTF_8.name(),
                        )
                }
            if (pairs.isNotEmpty()) {
                return requestUrl + (if (requestUrl.contains("?")) "&" else "?") + pairs.joinToString("&")
            }
        }
        return requestUrl
    }

    /** load() — đọc JS file từ plugin dir (có path-traversal guard, giới hạn 100KB). */
    private fun createLoadFunc(
        pluginSource: String?,
        extraScripts: Map<String, String>?,
        log: EngineLogger,
    ): BaseFunction = object : BaseFunction() {
        @Suppress("ReturnCount") // guard-chain validation trong BaseFunction override
        override fun call(
            cx: Context,
            scope: Scriptable,
            thisObj: Scriptable,
            args: Array<Any?>,
        ): Any? {
            val fileName = resolveLoadFileName(args)

            if (fileName.isEmpty()) {
                log.add("[load()] File name is empty")
                return null
            }
            if (!isValidFileName(fileName)) {
                log.add("[load()] Invalid file name: $fileName")
                return null
            }
            // In-memory trước — script src/*.js đã extract trong plugin zip
            // (load('config.js') resolve từ extraScripts, không cần disk)
            extraScripts?.get(fileName)?.let { content ->
                return evalExtraScript(cx, scope, fileName, content, log)
            }
            if (pluginSource.isNullOrEmpty()) {
                log.add("[load()] Plugin source directory not set")
                return null
            }

            return loadFromDisk(cx, scope, pluginSource, fileName, log)
        }

        override fun getDefaultValue(typeHint: Class<*>?): Any = "function load() { [native code] }"
    }

    /** Lấy file name từ arg đầu tiên của load(). */
    private fun resolveLoadFileName(args: Array<Any?>): String {
        if (args.isEmpty() || args[0] == null || args[0] === Undefined.instance) return ""
        return JsValueBridge.toJsString(args[0]) ?: ""
    }

    /** Eval script in-memory từ extraScripts; null khi lỗi eval. */
    @Suppress("LongParameterList") // tham số log DI + params nghiệp vụ Rhino
    private fun evalExtraScript(
        cx: Context,
        scope: Scriptable,
        fileName: String,
        content: String,
        log: EngineLogger,
    ): Any? = try {
        cx.evaluateString(scope, content, fileName, 1, null)
    } catch (e: Exception) {
        log.add("[load()] In-memory eval error $fileName: ${e.message}")
        null
    }

    /** Đọc + eval file từ đĩa (path-traversal guard, giới hạn 100KB); null khi lỗi. */
    @Suppress("ReturnCount", "LongParameterList") // security guards + tham số log DI — chain guard cần thoát sớm
    private fun loadFromDisk(
        cx: Context,
        scope: Scriptable,
        pluginSource: String,
        fileName: String,
        log: EngineLogger,
    ): Any? {
        return try {
            val file = File(pluginSource, "src/$fileName")
            if (!file.exists() || !file.isFile) {
                log.add("[load()] File not found: ${file.path}")
                return null
            }

            // Security: không thoát khỏi plugin dir
            val pluginDir = File(pluginSource, "src")
            if (!file.canonicalPath.startsWith(pluginDir.canonicalPath)) {
                log.add("[load()] Path traversal detected: $fileName")
                return null
            }

            // Security: giới hạn kích thước file
            if (file.length() > 100L * 1024) {
                log.add("[load()] File too large: ${file.length()} bytes")
                return null
            }

            val content = String(Files.readAllBytes(file.toPath()))
            cx.evaluateString(scope, content, fileName, 1, null)
        } catch (e: Exception) {
            log.add("[load()] Failed to load file: ${e.message}")
            null
        }
    }

    /** sleep(ms) — block thread hiện tại, interrupt-safe. */
    private fun createSleepFunc(): BaseFunction = object : BaseFunction() {
        override fun call(
            cx: Context,
            scope: Scriptable,
            thisObj: Scriptable,
            args: Array<Any?>,
        ): Any? {
            val millis = if (args.isNotEmpty() && args[0] is Number) (args[0] as Number).toInt() else 0
            return try {
                Thread.sleep(millis.toLong())
                null
            } catch (e: InterruptedException) {
                Thread.currentThread().interrupt()
                null
            }
        }

        override fun getDefaultValue(typeHint: Class<*>?): Any = "function sleep() { [native code] }"
    }

    /** Extract protocol://host[:port] từ URL đầy đủ. */
    fun extractBaseUrl(url: String): String = parseBaseUrl(url)

    /** Validate file name: không path traversal, chỉ .js, không ký tự lạ. */
    private fun isValidFileName(fileName: String): Boolean {
        if (fileName.isEmpty()) return false
        if (fileName.contains("..") || fileName.contains("/") || fileName.contains("\\")) return false
        if (fileName.startsWith("/") || fileName.startsWith("\\")) return false
        if (!fileName.endsWith(".js")) return false
        if (fileName.matches(Regex(".*[<>:\"|?*].*"))) return false
        return true
    }
}
